main/src/ca/uwaterloo/flix/language/errors/ParseError.scala

```diff
@@ -18,28 +18,234 @@ package ca.uwaterloo.flix.language.errors
-       |""".stripMargin
+    def message(formatter: Formatter): String = {
+      import formatter._
+      s"""${line(kind, source.name)}
+         |>> Expected ${red("scope")} on ${cyan(name)}.
+         |
+         |${code(loc, s"Here")}
+         |Hint: Add a scope using `@ <scope>`
+         |""".stripMargin
+    }
+  }
+
+  /**
+    * An error raised to indicate that one or more doc-comments are misplaced.
+    *
+    * @param ctx  The syntactic context.
+    * @param loc  The source location.
+    */
+  case class MisplacedDocComments(ctx: SyntacticContext, loc: SourceLocation) extends ParseError with Recoverable {
+    override val kind = s"Parse Error ($ctx)"
+    override val context: SyntacticContext = ctx
+
+    def summary: String = s"Misplaced doc-comment(s)."
+
+    def message(formatter: Formatter): String = {
+      import formatter._
+      s"""${line(kind, source.name)}
+         |>> Misplaced doc-comment(s).
+         |
+         |${code(loc, s"Here")}
+         |Hint: doc-comments must annotate declarations.
+         |""".stripMargin
+    }
+  }
+
+  /**
+    * An error raised to indicate that one or more sequential comments are misplaced.
+    *
+    * @param ctx  The syntactic context.
+    * @param loc  The source location.
+    */
+  case class MisplacedComments(ctx: SyntacticContext, loc: SourceLocation) extends ParseError with Recoverable {
+    override val kind = s"Parse Error ($ctx)"
+    override val context: SyntacticContext = ctx
+
+    def summary: String = s"Misplaced comment(s)."
+
+    def message(formatter: Formatter): String = {
+      import formatter._
+      s"""${line(kind, source.name)}
+         |>> Misplaced comment(s).
+         |
+         |${code(loc, s"Here")}
+         |Hint: Place comments on their own line.
+         |""".stripMargin
+    }
+  }
+
+  /**
+    * An error raised to indicate that something was syntactically malformed.
+    * This error is very generic, and should be avoided in favor of other ParserErrors if possible.
+    *
+    * @param name Name of the malformed item. See [[TokenKind.display]].
+    * @param ctx  The syntactic context.
+    * @param loc  The source location.
+    * @param hint Optional hint with more details about the error.
+    */
+  case class Malformed(name: String, ctx: SyntacticContext, loc: SourceLocation, hint: Option[String] = None) extends ParseError with Recoverable {
+    override val kind = s"Parse Error ($ctx)"
+    override val context: SyntacticContext = ctx
+
+    def summary: String = s"Malformed $name."
+
+    def message(formatter: Formatter): String = {
+      import formatter._
+      val hintStr = hint.map(s"\nHint: " + _).getOrElse("")
+      s"""${line(kind, source.name)}
+         |>> Malformed ${red(name)}.
+         |
+         |${code(loc, s"Here")}$hintStr
+         |""".stripMargin
+    }
+  }
+
+  /**
+    * An __internal__ error indicating that a Mark.Opened did not have a corresponding Mark.Closed when parsing ended.
```

Comment:
We want to guarantee that every `Mark.Opened` has a corresponding `Mark.Closed`. This is property is asserted in `buildTree`. But we also want each `Mark.Opened` to have a `TreeKind`. Before the mark is closed we _cannot_ know the kind, so instead we use `ErrorTree(UnclosedParserMark(...))`. This is safe because of the assert in `buildTree`. It's mostly a workaround to make types match, so I not particularly invested that things must be this way.

Another design could be to introduce a synthetic `TreeKind.UnclosedMark` with similar doc-comments. This would be another means to the same end, and be just as safe (because of the assert), but maybe we like the flavor of it better?

Reply:
Let me put it differently: Is this an error the programmer can encounter and fix?

Reply:
No.

Reply:
How/where do you discover this problem?

I think TreeKind.UnclosedMark would be better (?)

But maybe thats for another PR.

Reply:
On second thought, I'll include it here, since `ParseError.UnclosedParserMark` only exists on this branch.

Reply:
Why can't that assert throw an exception?